ipatests/test_integration/test_ipahealthcheck.py

```diff
@@ -904,3 +911,266 @@ def test_ipahealthcheck_trust_agent_member_check(self):
+            ]
+        )
+        yield
+        self.master.run_command(
+            [
+                "certutil",
+                "-d",
+                paths.ETC_DIRSRV_SLAPD_INSTANCE_TEMPLATE % instance,
+                "-A",
+                "-i",
+                paths.ETC_DIRSRV_SLAPD_INSTANCE_TEMPLATE % instance
+                + "/Server-Cert.pem",
+                "-t",
+                "u,u,u",
+                "-f",
+                paths.IPA_NSSDB_PWDFILE_TXT,
+                "-n",
+                "Server-Cert",
+            ]
+        )
+
+    def test_ipahealthcheck_ipansschainvalidation(self, remove_server_cert):
+        """
+        Test for IPANSSChainValidation
+        """
+        error_msg = (
+            ': certutil: could not find certificate named "Server-Cert": '
+            "PR_FILE_NOT_FOUND_ERROR: File not found\n"
+        )
+        returncode, data = run_healthcheck(
+            self.master, "ipahealthcheck.ipa.certs", "IPANSSChainValidation",
+        )
+        assert returncode == 1
+        for check in data:
+            if check["kw"]["nickname"] == "Server-Cert":
+                assert check["result"] == "ERROR"
+                assert check["kw"]["reason"] == error_msg
+
+    @pytest.fixture
+    def rename_raagent_cert(self):
+        """
+        Fixture to rename IPA RA CRT and revert
+        """
+        self.master.run_command(
+            ["mv", paths.RA_AGENT_PEM, "%s.old" % paths.RA_AGENT_PEM]
+        )
+        yield
+        self.master.run_command(
+            ["mv", "%s.old" % paths.RA_AGENT_PEM, paths.RA_AGENT_PEM]
+        )
+
+    def test_ipahealthcheck_iparaagent(self, rename_raagent_cert):
+        """
+        Testcase checks that ERROR message is displayed
+        when IPA RA crt file is renamed
+        """
+        error_msg = (
+            "[Errno 2] No such file or directory: '{}'"
+            .format(paths.RA_AGENT_PEM)
+        )
+        returncode, data = run_healthcheck(
+            self.master, "ipahealthcheck.ipa.certs", "IPARAAgent"
+        )
+        assert returncode == 1
+        for check in data:
+            assert check["result"] == "ERROR"
+            assert check["kw"]["error"] == error_msg
+
+    @pytest.fixture
+    def update_ra_cert_desc(self):
+        """
+        Fixture to modify description of RA cert in ldap
+        and revert
+        """
+        ldap = self.master.ldap_connect()
+        dn = DN(("uid", "ipara"), ("ou", "People"), ("o", "ipaca"))
+        entry = ldap.get_entry(dn)  # pylint: disable=no-member
+        ldap_cert_desc = entry.single_value.get("description")
+        desc = '200,12,CN=abc,CN=test'
+        entry.single_value['description'] = desc
+        ldap.update_entry(entry)  # pylint: disable=no-member
+        yield
+        ldap = self.master.ldap_connect()
+        dn = DN(("uid", "ipara"), ("ou", "People"), ("o", "ipaca"))
+        entry = ldap.get_entry(dn)  # pylint: disable=no-member
```

Comment:
You don't need to reconnect LDAP or re-do dn, they are still available from before the yield.

I think this can be expanded as a fixture into something more generic. Let me see what I can do, that way multiple errors can be detected using the same fixture.

Reply:
This will additionally test that a difference in just the serial # will pop up the right error:

```
diff --git a/ipatests/test_integration/test_ipahealthcheck.py b/ipatests/test_integration/test_ipahealthcheck.py
index 5bc473d63..7c1f7e8d0 100644
--- a/ipatests/test_integration/test_ipahealthcheck.py
+++ b/ipatests/test_integration/test_ipahealthcheck.py
@@ -1149,12 +1149,14 @@ class TestIpaHealthCheckWithExternalCA(IntegrationTest):
         dn = DN(("uid", "ipara"), ("ou", "People"), ("o", "ipaca"))
         entry = ldap.get_entry(dn)  # pylint: disable=no-member
         ldap_cert_desc = entry.single_value.get("description")
-        desc = '200,12,CN=abc,CN=test'
-        entry.single_value['description'] = desc
-        ldap.update_entry(entry)  # pylint: disable=no-member
-        yield
-        ldap = self.master.ldap_connect()
-        dn = DN(("uid", "ipara"), ("ou", "People"), ("o", "ipaca"))
+
+        def _update_entry(description):
+            entry = ldap.get_entry(dn)  # pylint: disable=no-member
+            entry.single_value['description'] = description
+            ldap.update_entry(entry)  # pylint: disable=no-member
+
+        yield _update_entry
+
         entry = ldap.get_entry(dn)  # pylint: disable=no-member
         entry.single_value['description'] = ldap_cert_desc
         ldap.update_entry(entry)  # pylint: disable=no-member
@@ -1165,6 +1167,26 @@ class TestIpaHealthCheckWithExternalCA(IntegrationTest):
         is modified, healthcheck tool reports the correct message
         """
         error_msg = 'RA agent not found in LDAP'
+        update_ra_cert_desc('200,12,CN=abc,CN=test')
+
+        returncode, data = run_healthcheck(
+            self.master,
+            "ipahealthcheck.ipa.certs",
+            "IPARAAgent",
+        )
+        assert returncode == 1
+        for check in data:
+            assert check["result"] == "ERROR"
+            assert check["kw"]["msg"] == error_msg
+
+    def test_ipahealthcheck_iparaagent_bad_serial(self, update_ra_cert_desc):
+        error_msg = 'RA agent description does not match. Found {got} ' \
+                    'in LDAP and expected {expected}'
+        update_ra_cert_desc(
+            '2;16;CN=Certificate Authority,O=%s;CN=IPA RA,O=%s' %
+            (self.master.domain.realm, self.master.domain.realm)
+        )
+
         returncode, data = run_healthcheck(
             self.master,
             "ipahealthcheck.ipa.certs",
@@ -1173,4 +1195,10 @@ class TestIpaHealthCheckWithExternalCA(IntegrationTest):
         assert returncode == 1
         for check in data:
             assert check["result"] == "ERROR"
+            assert check["kw"]["expected"] == '2;6;' \
+                'CN=Certificate Authority,O=%s;CN=IPA RA,' \
+                'O=%s' % (self.master.domain.realm, self.master.domain.realm)
+            assert check["kw"]["got"] == '2;16;' \
+                'CN=Certificate Authority,O=%s;CN=IPA RA,' \
+                'O=%s' % (self.master.domain.realm, self.master.domain.realm)
             assert check["kw"]["msg"] == error_msg
```